### ISSUE TITLE
[docs] Ember Data 1.13 to 2.0 Upgrade Guide, Step By Step

### ISSUE THREAD

**Issue by fivetanley** [emberjs/data#4145]
This issue will be used to solicit feedback on the upgrade woes people have had between 1.13 and 2.0. While the [Ember Data 1.13 blog post](http://emberjs.com/blog/2015/06/18/ember-data-1-13-released.html) does a great job of listing some changes, there's a good chance we didn't cover everything in the deprecation guide, and other bugs you have encountered.

Here are some questions I have for Ember Data users:
- What went well during the upgrade process?
- What did you find confusing about the Ember Data 1.13 blog post?
- What upgrade issues did we miss in the blog post?
- What addons did you need to upgrade to make compatible? Are there any addons that prevent you from upgrading?
- What would you like to see in an upgrade guide?
- Do you have a good example of step-by-step upgrade guides in other frameworks, such as Rails, Django, Spring, React, etc?
- What questions do you have after upgrading?

Thanks! I look forward to your responses.

**Comment by Gaurav0** [comment]
My previous response: https://github.com/emberjs/website/issues/2355

**Comment by workmanw** [comment]
Honestly the transition from Ember-data 1.0.0-Beta.19 to Ember 1.13 was the brutal one. Going from Ember Data 1.13 to 2.0 was pretty easy. If you're looking for feedback on the beta.19 transition I'd be happy to provide that.

**Comment by fivetanley** [comment]
@workmanw That would definitely be helpful.

**Comment by workmanw** [comment]
@fivetanley Sure. We did the migration to Ember 1.13 as soon as we could, so this is from recolation. 

> What went well during the upgrade process?

The RESTSerializer. Without that, transforming my REST payloads into JSON API payloads would have been terrible. 

> What did you find confusing about the Ember Data 1.13 blog post?

Well I felt like to fully understand the whole thing I had to go learn JSON API. It definitely took me some time to reorient my thinking around this new paradigm.

We have a custom serializer (and adapter) and it took me quite some time to figure out that the `RESTSerializer` had the utilities to convert my REST payloads to JSON API payloads. Also the way I was suppose to convert my serializer using `normalizeResponse` wasn't completely clear. 

> What upgrade issues did we miss in the blog post?

1) Issues with `pushPayload`. Prior to this upgrade we could do `store.pushPayload` and the store would accept a payload from our API. We did this A LOT with our realtime service. After this upgrade we aren't able to because `pushPayload` expected a JSON API formatted payload. I personally feel this was a break in the API, I opened an issued #3676 but it didn't get much action.

2) Changes to the error formatting. The blog post did document the changes to ` DS.InvalidError`, but I think it failed to properly highlight this change would impact the consumers. So for example: `user.errors.firstName` is now `user.errors.firstName.firstObject.message`. 

> What addons did you need to upgrade to make compatible? Are there any addons that prevent you from upgrading?

We did have to wait on ember-data-model-fragments to be compatible. This is our only ember-data based addon.

> What would you like to see in an upgrade guide?

I'm not really sure how to write a good upgrade guide ... but it definitely didn't feel like a step by step read. I had to read the whole thing twice and I didn't feel like I knew where I should start. Perhaps an upgrade guide would take you through it in more of a step by step fashion.

Also, I get the sense that a sizable minority of users have their own Adapters and Serializers. I feel that more attention could be payed to helping those users get migrated.

> Do you have a good example of step-by-step upgrade guides in other frameworks, such as Rails, Django, Spring, React, etc?

Sorry, nothing that i can think of.

> What questions do you have after upgrading?

One question I was really left with is why? What benefits are offered to users after this somewhat brutal upgrade. To me, surface of the store API is exactly the same before and after. The way I use ember-data is exactly the same. I spent 1 week and a half of time to upgrade and fix a bunch of tests and in the end not much changed. I'd really like it highlighted what I gained from my time investment. 

Secondarily, what's going on with `store.filter`. It's kind of in this addon, but it's really not the code still lives in the core.

**Comment by pwfisher** [comment]
Thank you, @workmanw, this is really helpful.

**Comment by acorncom** [comment]
> We have a custom serializer (and adapter) and it took me quite some time to figure out that the RESTSerializer had the utilities to convert my REST payloads to JSON API payloads. Also the way I was suppose to convert my serializer using normalizeResponse wasn't completely clear.

This was our biggest hang-up as well.  We made a couple of attempts at shifting to using the new Ember Data 1.13 API and had major snags (we were doing `pushPayload` as well).  It finally clicked on the second attempt that I could just adjust our JSON payload to match what the RESTSerializer was expecting and then have it handle the REST -> JSON-API conversion.  If that had been stated more clearly (or in the guides) it would have saved some headaches.  It's possible I missed it as well ;-)

Part of what made the second attempt easier was the guides had been updated with increased detail on serializers, which made it easier to see what we were trying to do: https://guides.emberjs.com/v2.3.0/models/customizing-serializers/

> What went well during the upgrade process?

We appreciated the `isNewSerializerAPI` approach quite a bit.  That made the transition far smoother than it otherwise would have been.  Kudos on handling things that way! :+1:
